File: examples/workflow-simple/src/main/java/org/axonframework/examples/workflow/simple/user/UserSignupWorkflow.java
/*
 * Copyright (c) 2010-2026. AxonIQ B.V.
 *
 * Licensed under the AXONIQ TERMS OF SERVICE,
 * Version 29 April 2026 (the "License");
 *
 * The software is available for evaluation use without registration.
 * Continued use beyond the evaluation period requires registration
 * and a commercial license. See the License for the specific language
 * governing permissions and limitations under the License.
 * You may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at:
 *  https://www.axoniq.io/legal/terms-of-service
 *
 * For licensing information and to register, visit:
 *  https://www.axoniq.io/pricing
 */
package org.axonframework.examples.workflow.simple.user;

import io.axoniq.framework.workflow.dsl.simple.SimpleWorkflowContext;
import io.axoniq.framework.workflow.runtime.api.annotation.Workflow;
import io.axoniq.framework.workflow.runtime.api.annotation.WorkflowCompletedHandler;
import io.axoniq.framework.workflow.runtime.api.execution.status.WorkflowStatus;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.time.Duration;
import java.time.Instant;
import java.util.Map;

import static io.axoniq.framework.workflow.runtime.association.Associations.associate;
import static io.axoniq.framework.workflow.dsl.api.Payload.payload;
import static io.axoniq.framework.workflow.dsl.base.BaseWorkflowContext.equalsTo;
import static io.axoniq.framework.workflow.runtime.association.PayloadPropertyValueRetriever.payloadProperty;

/**
 * Sample user registration.
 * @since 5.4.0
 */
public class UserSignupWorkflow {

    Logger logger = LoggerFactory.getLogger(UserSignupWorkflow.class);

    @Workflow(
            idProperty = "id",
            startOnEventClass = RegistrationReceivedEvent.class,
            startOnConditions = {"payload:status=vip"},
            workflowName = "MyWorkflow",
            workflowVersion = "1.0.0",
            workflowNamespace = "io.axoniq.dsl.wf.workflow"
    )
    public void execute(
            SimpleWorkflowContext ctx
    ) {

        logger.info("User signup workflow started at {} for {}", Instant.now(), ctx.workflowPayload());

        // -> start
        var success = ctx.awaitExecute("createUser", Boolean.class, UserService::createUser);
        if (!success) {
            return;
        }

        ctx.awaitExecute("activateUser",
                         ctx.workflowPayload(),
                         UserService::activateUser,
                         step -> step.timeout(Duration.ofSeconds(10))
        );

//        ctx.awaitExecute("activateUser2",
//                         ctx.workflowPayload(),
//                         UserService::activateUser,
//                         step -> step.timeout(Duration.ofSeconds(10))
//        );

        ctx.awaitExecute("sendWelcomeEmail",
                         ctx.workflowPayload(),
                         (pc, input) -> {
                             NotificationService.sendEmail(payload(input).get("email"));
                             return Map.of();
                         });
        ctx.sleep("waitASecond", Duration.ofSeconds(1L));

        var magic = ctx.awaitEvent("waitForMagicToHappen",
                                   MagicHappenedEvent.class,
                                   associate(payloadProperty("magician"), equalsTo("Merlin")),
                                   step -> step.timeout(Duration.ofSeconds(5))
        );
        ctx.setPayload("modifyPayload", magic);

        logger.info("Magic happened because of the magician {}", magic.magician());
        // -> end
    }

    @WorkflowCompletedHandler
    public void onFinish(
            WorkflowStatus workflowStatus,
            SimpleWorkflowContext ctx
    ) {
        logger.info("User signup workflow {} at {} for {}", workflowStatus, Instant.now(), ctx.workflowPayload());
    }
}

File: examples/workflow-simple/src/main/kotlin/org/axonframework/examples/workflow/simple/kotlin/user/UserSignupWorkflow.kt
/*
 * Copyright (c) 2010-2026. AxonIQ B.V.
 *
 * Licensed under the AXONIQ TERMS OF SERVICE,
 * Version 29 April 2026 (the "License");
 *
 * The software is available for evaluation use without registration.
 * Continued use beyond the evaluation period requires registration
 * and a commercial license. See the License for the specific language
 * governing permissions and limitations under the License.
 * You may not use this file except in compliance with the License.
 *
 * You may obtain a copy of the License at:
 *  https://www.axoniq.io/legal/terms-of-service
 *
 * For licensing information and to register, visit:
 *  https://www.axoniq.io/pricing
 */
package org.axonframework.examples.workflow.simple.kotlin.user

import org.axonframework.examples.workflow.simple.kotlin.user.MagicHappenedEvent
import org.axonframework.examples.workflow.simple.kotlin.user.NotificationService
import org.axonframework.examples.workflow.simple.kotlin.user.UserService
import io.axoniq.framework.workflow.dsl.kotlin.Kontext
import io.axoniq.framework.workflow.runtime.api.annotation.Workflow
import io.axoniq.framework.workflow.runtime.api.execution.context.EventConditions
import io.axoniq.framework.workflow.runtime.api.execution.context.retry.RetryPolicy
import io.github.oshai.kotlinlogging.KotlinLogging
import java.time.Instant
import kotlin.time.Duration.Companion.seconds


private val logger = KotlinLogging.logger {}

/**
 * @since 5.4.0
 */
class UserSignupWorkflow {

    @Workflow(idProperty = "id", startOnEventName = "my.custom.RegistrationReceived")
    fun Kontext.onExecute() {
        logger.info { "User signup workflow started at ${Instant.now()} for $payload" }

        val success = awaitExecute<Boolean>(
            "createUser",
            timeout = 5.seconds,
            retryPolicy = RetryPolicy.NONE
        ) { UserService.createUser() }

        if (!success) {
            return
        }

        block {
            execute(
                stepName = "activateUser",
                inputPayload = payload,
                timeout = 10.seconds,
                retryPolicy = RetryPolicy.NONE,
                action = UserService::activateUser
            )
        }

        val waitASecond = waitForEvent("waitASecond", EventConditions.never(), timeout = 1.seconds)
        waitASecond.await()
        if (waitASecond.failure() && waitASecond.error().isPresent) {
            throw waitASecond.error().get()
        }

        block {
            execute(
                stepName = "sendWelcomeEmail",
                inputPayload = payload,
                timeout = 5.seconds,
                retryPolicy = RetryPolicy.NONE
            ) { _, _ ->
                NotificationService.sendEmail()
                mapOf()
            }
        }

        val magic = awaitEvent<MagicHappenedEvent>(
            "waitForMagicToHappen",
            timeout = 5.seconds
        )

        logger.info { "Magic happened because of the magician ${magic.magician}" }
        logger.info { "User signup workflow ended at ${Instant.now()} for $payload" }
    }

}

fun UserSignupWorkflow.execute(kontext: Kontext) = with(kontext) { onExecute() }


